src/Grpc.AspNetCore.Web/Internal/Base64PipeReader.cs

```diff
@@ -0,0 +1,193 @@
+        {
+            _inner = inner;
+        }
+
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            var consumedPosition = ResolvePosition(consumed);
+
+            _inner.AdvanceTo(consumedPosition);
+        }
+
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            var consumedPosition = ResolvePosition(consumed);
+            var examinedPosition = ResolvePosition(examined);
+
+            _inner.AdvanceTo(consumedPosition, examinedPosition);
+        }
+
+        private SequencePosition ResolvePosition(SequencePosition base64Position)
+        {
+            // TODO - is there a better way to get the index of a position in a sequence?
+            var length = _currentDecodedBuffer.Slice(0, base64Position).Length;
+
+            // IMPORTANT - this logic assumes that SequencePosition will be multiples of 3.
+            // GetMaxEncodedToUtf8Length will round up to next multiple of 3, so more data
+            // could be advanced internally than expected if not called using a mulitple of 3.
+            //
+            // Because ReadAsync always returns data in multiples of 3, and gRPC only uses
+            // ReadResult.Buffer.End to advance, the position should always be a multiple
+            // of 3. The only time it won't be is at the end of a message, in which case
+            // rounding up is the correct thing to do to consume padding.
+            var resolvedPosition = _currentInnerBuffer.GetPosition(Base64.GetMaxEncodedToUtf8Length((int)length));
+            return resolvedPosition;
+        }
+
+        public override void CancelPendingRead()
+        {
+            _inner.CancelPendingRead();
+        }
+
+        public override void Complete(Exception? exception = null)
+        {
+            _inner.Complete(exception);
+        }
+
+        public async override ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            var innerResult = await _inner.ReadAsync(cancellationToken);
+            if (innerResult.Buffer.IsEmpty)
+            {
+                _currentDecodedBuffer = innerResult.Buffer;
+                _currentInnerBuffer = innerResult.Buffer;
+                return innerResult;
+            }
+
+            // Minimum valid base64 length is 4. Read until we have at least that much content
+            while (innerResult.Buffer.Length < 4)
+            {
+                if (innerResult.IsCompleted)
+                {
+                    // If the reader completes with less than 4 bytes then the base64 isn't valid..
+                    throw new InvalidOperationException("Unexpected end of data when reading base64 content.");
+                }
+
+                if (innerResult.IsCanceled)
+                {
+                    return new ReadResult(
+                        ReadOnlySequence<byte>.Empty,
```

Comment:
Are you sure you can advance after returning this canceled ReadResult? I think this would result in unresolvable SequencePositions being passed to the next call to AdvanceTo. It's possible that _currentDecodedBuffer and _currentInnerBuffer might not even be set which would lead to a null ref.

Reply:
Done.

btw I experimented and PipeReader.Advance isn't required after a canceled result, but you can choose to do it. I fixed some issues so CancelPendingReader -> Advance this can succeed.